In [1]:
import numpy as np

In [2]:
X = np.array([[5, 9], [-2, 3], [4, 8], [3, -2], [-3, -7]])
y = np.array([16, 1, 14, 21, 8])

In [3]:
def predict(w):
    ŷ = []
    for i in range(len(X)):
        p̂ = w[0] * X[i][0] + w[1] * X[i][1] + w[2]
        ŷ.append(p̂)
    return ŷ

In [4]:
def loss(w):
    ŷ = predict(w)
    mse = np.mean((y - ŷ) ** 2)
    rmse = np.sqrt(mse)
    return rmse

In [5]:
def gradient(w, 𝛿 = 0.001):
    g = []
    loss_0 = loss(w)
    for i in range(3):
        weights = np.copy(w)
        weights[i] += 𝛿
        loss_1 = loss(weights)
        gradient = (loss_1 - loss_0) / 𝛿
        g.append(gradient)
    return g

In [6]:
def update(w, n = 10):
    for i in range(3):
        w[i] -= n * gradient(w)[i]
    return w

In [7]:
#First set of weights
w = [1, 1, 1]
initial_loss = loss(w)
updated_w = update(w)
new_loss = loss(updated_w)

print(f'Initial Loss: {initial_loss}')
print(f'Update Weights: {updated_w}')
print(f'New Loss: {new_loss}')

Initial Loss: 11.428035701729321
Update Weights: [1.0, -24.04154287978011, 4.979486314733549]
New Loss: 155.49198683927918


In [8]:
#Second set of weights
w = [-1, -1, 0]
initial_loss = loss(w)
updated_w = update(w)
new_loss = loss(updated_w)

print(f'Initial Loss: {initial_loss}')
print(f'Updated Weights: {updated_w}')
print(f'New Loss: {new_loss}')

Initial Loss: 20.33715811021786
Updated Weights: [30917.6889837801, -49.455247888341546, -3.941996692446992]
New Loss: 109496.16339013436


# Did the loss get lower after the parameters were updated? Does this mean the code is working properly?

**First set of Weights:** 
- Initial Loss: 11.428
- New Loss: 10.850

The loss decreased after updating the parameters. This suggests that the code is working properly, as the level of error decreased, indicating that the model is reliable.


**Second set of Weights:**
- Initial Loss: 20.337
- New Loss: 1077.043

After updating the parameters, the loss increased significantly. This suggests that the code might not be working correctly, as the level of error increased, which tells us that the model is unreliable.

# Did the loss get lower this time? Explain what went wrong and why this is bad.

**First set of Weights:**
- Initial Loss: 11.428
- New Loss: 155.492

**Second set of Weights:**
- Initial Loss: 20.337
- New Loss: 109496.163

**Explanation:**
The loss increased significantly after setting the learning rate to 10. A higher learning rate can cause the model to overshoot the minimum, leading to higher error values. This is bad because it prevents the model from converging to the optimal solution, making it unreliable.